### PR TITLE
Port table semtype to jballerina

### LINKED ISSUE #34284: Port table semtype to jballerina
**Description:**
Port changes done to nballerina in https://github.com/ballerina-platform/nballerina/pull/747

**Describe your task(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by ushirask**
Fixes #34284

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -644,6 +646,13 @@ private SemType resolveTypeDesc(BLangBuiltInRefTypeNode td, Env semtypeEnv) {
         }
     }
 
+    private SemType resolveTypeDesc(BLangTableTypeNode td, Env semtypeEnv, Map<String, BLangNode> mod, int depth,
+                                    BLangTypeDefinition typeDefinition) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:650]: Unused parameter `typeDefinition`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -644,6 +646,13 @@ private SemType resolveTypeDesc(BLangBuiltInRefTypeNode td, Env semtypeEnv) {
...
+        SemType memberType = resolveTypeDesc(semtypeEnv, mod, (BLangTypeDefinition) td.constraint.defn, depth,
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:651]: Can we break just after `=` sign to get all in same line, or else I think we can try to break just after the open parenthesis.

File: semtypes/src/main/java/io/ballerina/types/OpsTable.java [hunk 3]
```diff
@@ -64,7 +65,7 @@
         OPS[i++] = PANIC_IMPL;          // RW stream
         OPS[i++] = new ListTypeRwOps(); // RW list
         OPS[i++] = new MappingRwOps();  // RW mapping
-        OPS[i++] = PANIC_IMPL;          // RW table
+        OPS[i++] = new TableRwOps();  // RW table
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/types/OpsTable.java:68]: shall we push the comment by one more space to align with the rest

File: semtypes/src/main/java/io/ballerina/types/subtypedata/RwTableSubtype.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import io.ballerina.types.Bdd;
+import io.ballerina.types.ProperSubtypeData;
+
+/**
+ * Contains 2 bdds for readonly and readwrite mappings.
+ *
+ * @since 3.0.0
+ */
+public class RwTableSubtype implements ProperSubtypeData {
+    public Bdd ro;
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/types/subtypedata/RwTableSubtype.java:29]: Shall we make these final

File: semtypes/src/main/java/io/ballerina/types/subtypedata/RwTableSubtype.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    public Bdd rw;
+
+    private RwTableSubtype(Bdd ro, Bdd rw) {
+        this.ro = ro;
+        this.rw = rw;
+    }
+
+    public static ProperSubtypeData createRwTableSubtype(Bdd ro, Bdd rw)  {
+        return new RwTableSubtype(ro, rw);
```
Comment (rdhananjaya) [inline semtypes/src/main/java/io/ballerina/types/subtypedata/RwTableSubtype.java:38]: Don't we need to port this bit from ballerina code 
```ballerina
    if ro is boolean && ro == rw {
        return ro;
    }
```

Comment (ushirask) [reply semtypes/src/main/java/io/ballerina/types/subtypedata/RwTableSubtype.java:42]: Fixed it, thanks.

**Pull request merged**